std/stdio.d

```diff
@@ -1846,14 +1846,16 @@ is recommended if you want to process a complete file.
         }
         else
         {
-            // TODO: optimize this
             string s = readln(terminator);
-            buf.length = 0;
-            if (!s.length) return 0;
-            foreach (C c; s)
-            {
-                buf ~= c;
+            if (!s.length) {
+                buf = buf[0..0];
+                return 0;
             }
+
+            buf.length = s.length;
+            foreach (i, C c; s)
+                buf[i] = c;
```

Comment:
You need to use appending. This is the trans-coding path, which means you are reading `dchar` or `wchar` instead of char.

Using an appender might work, or you can use:
```d
buf = buf[0 .. 0];
buf.assumeSafeAppend;
// append same as before
```

Reply:
Didn't knew the `.assumeSafeAppend`. Interesting, but isn't this unsafe? Also, I need to use
```d
import std.utf : codeLength;
buf.length = codeLength!C(s);
```
To determinate the size.

Also, with the following code, the performance is worst than using an appender:
```d
string s = readln(terminator);
if (!s.length)
{
    buf = buf[0..0];
    return 0;
}
import std.utf : codeLength;
buf.length = codeLength!C(s);
buf = buf[0..0];
foreach (i, C c; s)
    buf.assumeSafeAppend() ~= c;
return buf.length;
```

![image](https://user-images.githubusercontent.com/5050669/97353356-dc459300-188b-11eb-8e4a-aaf7f44cbcbf.png)
Code: https://gist.github.com/run-dlang/8e6e06e357cf4761cc5000d24fdfa4e6

Reply:
No just call `assumeSafeAppend` once. And it's probably going to be worse than appender, because appender doesn't need to go through opaque calls to figure out whether a block has space.

But yes, you could use codeLength to make it work, and just use an incrementing index.

Reply:
It's not unsafe, as long as you are just writing characters, and you aren't overwriting immutable data.

Reply:
> No just call `assumeSafeAppend` once. And it's probably going to be worse than appender, because appender doesn't need to go through opaque calls to figure out whether a block has space.

Yes, it got better but worse than using an appender.

> But yes, you could use codeLength to make it work, and just use an incrementing index.

I thought that index provided by `_aApplycw2` was incremental but isn't suitable for this case because it does `decode(buf, i)` instead.

My conclusion is that using incrementing index and resizing with codeLength is the best way to go.

Order:
1. Old implementation
2. Appender
3. assumeSafeAppend + codeLength
4. incremental idx + codeLength
![image](https://user-images.githubusercontent.com/5050669/97357309-765c0a00-1891-11eb-9fbc-43fcec075532.png)